src/runtime/recipe/recipe.ts

```diff
@@ -627,8 +628,20 @@ export class Recipe {
     return this.handles.filter(handle => handle.connections.length === 0);
   }
 
-  getFreeConnections() {
-    return this.handleConnections.filter(hc => !hc.handle && !hc.isOptional);
+  getFreeConnections(type?: Type): {particle: Particle, connSpec: ConnectionSpec}[] {
+    const freeConnections: {particle: Particle, connSpec: ConnectionSpec}[] = [];
+    for (const particle of this.particles) {
+      if (particle.spec) {
+        for (const connSpec of particle.spec.connections) {
```

Comment:
if you add a specifiedConnections method to Recipe that grabs particle.spec.connections off all the particles, then you can go back to a filter operation here and below, which is a lot simpler and easier to read.

Reply:
done - and got rid from the duplicate method, this is much better now, thanks!